presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/WeightedDoubleReservoirSample.java

```diff
@@ -0,0 +1,240 @@
+            }
+        }
+    }
+
+    private void swap(int fpos, int spos)
+    {
+        double tmpElement = samples[fpos];
+        double tmpWeight = weights[fpos];
+        samples[fpos] = samples[spos];
+        weights[fpos] = weights[spos];
+        samples[spos] = tmpElement;
+        weights[spos] = tmpWeight;
+    }
+
+    private void bubbleDown()
+    {
+        int index = 0;
+        while (leftChild(index) < size) {
+            int smallestChildIndex = leftChild(index);
+
+            if (rightChild(index) < size && weights[leftChild(index)] > weights[rightChild(index)]) {
+                smallestChildIndex = rightChild(index);
+            }
+
+            if (weights[index] > weights[smallestChildIndex]) {
+                swap(index, smallestChildIndex);
+            }
+            else {
+                break;
+            }
+
+            index = smallestChildIndex;
+        }
+    }
+
+    private void bubbleUp()
+    {
+        int index = size - 1;
+        while (index > 0 && weights[index] < weights[parent(index)]) {
+            swap(index, parent(index));
+            index = parent(index);
+        }
+    }
+
+    private static int parent(int pos)
+    {
+        return pos / 2;
+    }
+
+    private static int leftChild(int pos)
+    {
+        return 2 * pos;
+    }
+
+    private static int rightChild(int pos)
+    {
+        return 2 * pos + 1;
+    }
+
+    public static WeightedDoubleReservoirSample deserialize(SliceInput input)
+    {
+        int size = input.readInt();
+        int maxSamples = input.readInt();
+        checkArgument(size <= maxSamples, "size must not be larger than number of samples");
+        double[] samples = new double[maxSamples];
+        input.readBytes(
+                Slices.wrappedDoubleArray(samples),
+                size * SizeOf.SIZE_OF_DOUBLE);
+        double[] weights = new double[maxSamples];
+        input.readBytes(
+                Slices.wrappedDoubleArray(weights),
+                size * SizeOf.SIZE_OF_DOUBLE);
+        return new WeightedDoubleReservoirSample(size, samples, weights);
+    }
+
+    public void serialize(SliceOutput out)
+    {
+        out.appendInt(size);
+        out.appendInt(samples.length);
+        IntStream.range(0, size).forEach(i -> out.appendDouble(samples[i]));
+        IntStream.range(0, size).forEach(i -> out.appendDouble(weights[i]));
+    }
+
+    public int getRequiredBytesForSerialization()
+    {
+        return SizeOf.SIZE_OF_INT + // size
+                SizeOf.SIZE_OF_INT + // maxSamples
+                2 * SizeOf.SIZE_OF_DOUBLE * Math.min(size, samples.length); // samples, weights
+    }
+
+    public long estimatedInMemorySize()
+    {
+        return INSTANCE_SIZE +
+                SizeOf.SIZE_OF_INT + // size
+                SizeOf.SIZE_OF_INT + // maxSamples
+                2 * SizeOf.SIZE_OF_DOUBLE * Math.min(size, samples.length); // samples, weights
```

Comment:
This should be `INSTANCE_SIZE + sizeOf(samples) + sizeOf(weights)`

Reply:
It depends on whether it's the estimated size for memory, or serialization size, no? I can't tell anymore to which one does the comment here refer.

Reply:
@atavory This is for in-memory size.

Reply:
Changed the code in the serialization one, and kept the one here for consistency with that one.

Reply:
@atavory These two methods should not be consistent. They report different things. One reports number of bytes this object will use when serialized. The other reports number of bytes used to store this object in JVM's memory.

Reply:
I was just referring to the style, not the return value.